common/cuda_hip/matrix/bccoo_helper.hpp.inc

```diff
@@ -0,0 +1,278 @@
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+namespace {
+
+
+// Routines for block compression objects
+
+
+template <int subwarp_size = config::warp_size, typename IndexType,
+          typename ValueType, typename Closure>
+inline __device__ void loop_block_single_row(
+    const uint8* __restrict__ chunk_data, size_type block_size_local,
+    const ValueType* __restrict__ b, const size_type b_stride,
+    const size_type column_id, ValueType* __restrict__ c,
+    const size_type c_stride, compr_idxs& idxs, compr_blk_idxs& blk_idxs,
+    const size_type start_in_blk, const size_type jump_in_blk, Closure scale)
+{
+    ValueType temp_val = zero<ValueType>();
+    bool new_value = false;
+    ValueType val;
+    const auto tile_block =
+        group::tiled_partition<subwarp_size>(group::this_thread_block());
+
+
+    for (size_type pos = start_in_blk; pos < block_size_local;
+         pos += jump_in_blk) {
+        //        if (pos < block_size_local) {
+        idxs.col = blk_idxs.col_frs +
+                   get_value_chunk<IndexType>(
+                       chunk_data, blk_idxs.shf_col + pos * sizeof(IndexType));
+        val = get_value_chunk<ValueType>(
+            chunk_data, blk_idxs.shf_val + pos * sizeof(ValueType));
+        temp_val += val * b[idxs.col * b_stride + column_id];
+        //        }
+        new_value = true;
+    }
+    if (new_value) {
```

Comment:
will all threads of a subwarp use the above for loop? if not, the condition should use `tile_block.any(new_value)` to ensure all threads in subwarp to use segment_scan

Reply:
True!!